xarray/core/dataset.py

```diff
@@ -1212,7 +1212,18 @@ def _construct_dataarray(self, name: Hashable) -> DataArray:
         coords: dict[Hashable, Variable] = {}
         # preserve ordering
         for k in self._variables:
-            if k in self._coord_names and set(self._variables[k].dims) <= needed_dims:
+            var_dims = set(self._variables[k].dims)
+            if k in self._indexes:
+                try:
+                    self._indexes[k].validate_dataarray_coord(
+                        k, self._variables[k], needed_dims
+                    )
+                    coords[k] = self._variables[k]
+                except ValueError:
```

Comment:
`ValueError` can be raised by many different things going wrong in Python. If we're going to check specifically for an error, we should only catch a specific Exception subclass.

Reply:
Added `CoordinateValidationError` in [5da014e](https://github.com/pydata/xarray/pull/10137/commits/5da014e260576b8c0579b312e0bde42a6461e5eb).

For now it is defined in `xarray/core/indexes.py` to avoid a cyclic import. That's a bit messy but we could maybe move the definition of all Xarray custom error types in a dedicated `xarray.exception` module (in a later PR?).